### ISSUE TITLE
Question: recommended way to package native dependencies

### ISSUE THREAD

**Issue by psantosl** [dotnet/runtime#32161]
Hi,

Hopefully this is a right place to ask questions, not sure if I have to tag them somehow.

We are about to replace our Mono-based **Linux** server with .NET Core officially on production (we've been tinkering with it for years already).

Right now:

- We build or own Mono (for Linux).
- Package it for all our supported distros together with our .NET assemblies using [OBS](https://build.opensuse.org/).
- Set the dependencies with libraries we use.
- Include the native libraries we build.

In .NET Core:
- We use things like libz or liblz4. **What's the best way to package them together with the bundle?** I mean, yes, we can simply copy them to the output path, I mean if there's something better/officially supported.
- Same if we want to include them in a **PublishSingleFile** deployment.
- And finally, what's the recommended way to deploy the [.NET Core runtime dependencies](https://github.com/dotnet/core/blob/master/Documentation/prereqs.md) themselves? Add the rpms or something to our OBS? (We dream with getting rid of OBS and just deploy a single native build with CoreRT, but that's another story).

Thanks!

**Comment by Clockwork-Muse** [comment]
> . We use things like libz or liblz4. What's the best way to package them together with the bundle? I mean, yes, we can simply copy them to the output path, I mean if there's something better/officially supported.
> . Same if we want to include them in a PublishSingleFile deployment.

It's possible to create NuGet packages from native binaries (check if the ones you use already exist - otherwise, set up a private NuGet server).  This would mean that from your application point of view, the included dll/so/whatever is just another artifact, so just gets included "normally".  You can package multiple deploy targets in the same package, so you don't need separate packages for each one either.

> . And finally, what's the recommended way to deploy the .NET Core runtime dependencies themselves? Add the rpms or something to our OBS? (We dream with getting rid of OBS and just deploy a single native build with CoreRT, but that's another story).

If you're running a server, the thing to do would be to place them in a docker container, no?  [There are official base images](https://hub.docker.com/_/microsoft-dotnet-core/), and (especially for ASP.NET Core) there are existing templates to handle this.  This means that your deployment is then a normal docker container, which you can then handle through Kubernetes, or whatever.

**Comment by jkotas** [comment]
>  I mean, yes, we can simply copy them to the output path

If you just need them for your app and you do not intend to distribute them as NuGet packages, simply copying them to the output path is a fine solution.

> include them in a PublishSingleFile deployment.

@swaroop-sridhar Could you please comment on this?

> what's the recommended way to deploy the .NET Core runtime dependencies

It depend on what kind of installer you have for your product. If it is delivered as container, it is easy as @Clockwork-Muse pointed out. If it is delivered via OS package manager (e.g. `apt-get`), the dependencies needs to be understood by the package manager.

**Comment by swaroop-sridhar** [comment]
> We use things like libz or liblz4. What's the best way to package them together with the bundle? 
> Same if we want to include them in a PublishSingleFile deployment.

@psantosl how are the dependencies obtained? The `dotnet publish` command picks up dependencies (from nuget / dependent projects, etc) and places them in the publish directory. 

Once this is done, publishing to a single-file (by setting the `PublishSingleFile` property to `true`) will package the published binaries (managed or native) to a single file.

**Comment by am11** [comment]
related: https://github.com/dotnet/standard/issues/584

**Comment by psantosl** [comment]
Thanks everyone!

> @psantosl how are the dependencies obtained? The dotnet publish command picks up dependencies (from nuget / dependent projects, etc) and places them in the publish directory.

Right now we build some of the dependencies with OBS, I mean, we put everything into a directory and then package it. But I'm talking about pre-NET Core, this is how we do it for Mono.

**Comment by psantosl** [comment]
> If you're running a server, the thing to do would be to place them in a docker container, no?

Well, that's something to look into @Clockwork-Muse 👍 Although my dream is closer to single binary with CoreRT... maybe I'm too old-school :-/

**Comment by Clockwork-Muse** [comment]
Even for standalone binaries software deployment is moving to some form of containerization - [`snap`s on Linux, for example](https://snapcraft.io/).  Because dependency management is an absolute pain generally.

One of the big benefits to moving to a docker container, though, would be the fact that it becomes your run target for developers.  Literally no more "but it runs on my machine".

**Comment by joperezr** [comment]
closing as it seems like the original question was answered. Feel free to reopen if you think otherwise.
